docs/markdown/Syntax.md

```diff
@@ -527,6 +527,54 @@ executable('progname', 'prog.c',
 
 Attempting to do this causes Meson to immediately exit with an error.
 
+### Argument flattening
+
+Argument flattening is a Meson feature that aims to simplify using
+methods and functions. For functions where this feature is active,
+Meson takes the list of arguments and flattens all nested lists into
+one big list.
+
+For instance the following function calls to [[executable]] are
+identical in Meson:
+
+```meson
+# A normal example:
+executable('exe1', ['foo.c', 'bar.c', 'foobar.c'])
+
+# A more contrived example that also works but certainly
+# isn't good Meson code:
+l1 = ['bar.c']
+executable('exe1', [[['foo.c', l1]], ['foobar.c']])
+
+# How meson will treat all the previous calls internally:
+executable('exe1', 'foo.c', 'bar.c', 'foobar.c')
+```
+
+Because of an internal implementation detail, the following syntax
```

Comment:
Hm, I didn't know about this implementation detail. Why are we documenting it? Shouldn't we detect it and print a warning on usage instead?

Reply:
I think the problem is that the current interpreter implementation also makes it effectively very hard to know that that is happening?

Reply:
I have already tried to detect / properly deprecate this, but it is nearly impossible with the current interpreter setup. It will be possible once we have every function (that supports argument flattening) type-checked with `@typed_pos_args` and friends. In that case, we could (and should) move the logic to there.

My point for documenting this is that we can't really deprecate it because we currently can't feasibly check it. So being transparent about this fact (and telling people that this code *will* break in the future) in the docs is the best solution I could come up with.

Reply:
I don't see the point of documenting it in such detail either way. We don't generally put bugs / unintentional behaviour in documentation unless users are very likely to encounter it. What am I missing?

Reply:
IMHO, it is not unreasonable to unintentionally run into this if you do something like this:

```meson
src = ['main', 'main.cpp']

executable(src)
```

So my point is that this is the best solution I could currently come up with to warn users **not** to do this.